### ISSUE TITLE
test_get_candidates failed on non-FHS system (NixOS)

### ISSUE THREAD

**Issue by definfo** [wlh320/rime-ls#32]
I tried to build it on NixOS. Compilation passed, but test_get_candidates failed.
Then I found the related code in `src/rime.rs`.
```rust
#[test]
fn test_get_candidates() {
    let shared_data_dir = "/usr/share/rime-data/";
```
This assumes a fixed path for shared data. However, NixOS does not provide such path since it's non-FHS.

**Comment by wlh320** [comment]
Actually, not just NixOS, but most systems should not be able to pass this test, because this hard-coded path for the Rime shared data seems to only work on Linux. On NixOS, where is the shared data for  Rime typically stored? Maybe it can be improved by conditional compilation?

**Comment by definfo** [comment]
> On NixOS, where is the shared data for Rime typically stored?

In convention, `/nix/store/<hash>-rime-xxx-<version>/share/rime-data/`
Notice that /nix/store is read-only to ensure reproducibility.

A common practice is to add a rime data symlink, see https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/tools/inputmethods/fcitx5/fcitx5-rime.nix#L42
TLDR: add rime-data as input, and move it into the final package during postInstall phase.

**Comment by wlh320** [comment]
I tried to fix it with a helper function, which reads environment variable `RIME_DEFAULT_DATA_DIR` at compile time, if not set, it uses known default shared data dir on different platforms. It  works not only for cargo test, but also for default LSP initialization options.

**Comment by wlh320** [comment]
I noticed that there already been a environment variable in Makefile of librime for rime shared data, so I changed the name to keep the same as it.  See [5037bb6](https://github.com/wlh320/rime-ls/commit/5037bb6dc91f71447520d9e6ac09bb95410b131d)

**Comment by definfo** [comment]
Thanks, now all tests pass.
My package definition is here https://github.com/definfo/nur-packages/blob/master/pkgs/rime-ls/default.nix .
Maybe I can open a PR to nixpkgs right after the next release.
As for editor/IDE configuration, I will try adding some script for assistance, since `shared_data_dir` on NixOS will change on package upgrade.

By the way, could you add `Cargo.lock` to your repo? Lockfile is crucial for reproductive build, and Nix package definition relies on it. I work around by generating it manually through `cargo fetch` and patch into the repo directory.

**Comment by wlh320** [comment]
Thanks. I used to directly use `.gitignore` generated by Cargo, but it was only after your reminder that I realized it recommends removing `Cargo.lock` from gitignore if creating an executable.

**Comment by neo451** [comment]
@wlh320 @definfo I am new to nixos, and want to setup rime-ls again on this machine, I did everything right, except I think non-FHS could still be the reason why it is not working, I don't want to do everything in nix, and is not using nur, I just want to set the correct shared data dir in my lua config, would really appreciate some help :)

**Comment by definfo** [comment]
> [@wlh320](https://github.com/wlh320) [@definfo](https://github.com/definfo) I am new to nixos, and want to setup rime-ls again on this machine, I did everything right, except I think non-FHS could still be the reason why it is not working, I don't want to do everything in nix, and is not using nur, I just want to set the correct shared data dir in my lua config, would really appreciate some help :)

Sorry I didn't update the manual for NixOS. I will update here ASAP :)

**Comment by neo451** [comment]
> > [@wlh320](https://github.com/wlh320) [@definfo](https://github.com/definfo) I am new to nixos, and want to setup rime-ls again on this machine, I did everything right, except I think non-FHS could still be the reason why it is not working, I don't want to do everything in nix, and is not using nur, I just want to set the correct shared data dir in my lua config, would really appreciate some help :)
> 
> Sorry I didn't update the manual for NixOS. I will update here ASAP :)

wow thank you for your quick response
